docs/deployment-ux.md

```diff
@@ -0,0 +1,410 @@
+SomeEnvironment/Prod:2    Active       6         6         6            6        
+```
+
+### Daemon update semantics
+
+Daemon deployments have two update methods, with different availability
+impacts
+
+1.  Terminate before replace
+
+    If you're using this deployment method, then Blox will kill
+    `(100 - MinHealthyPercent)%` of tasks, and only launch new tasks to
+    replace as they are successfully terminated. This will ensure that
+    there's never more than one version of the Task running at the same
+    time on the same host, but may result in some downtime for the
+    daemon. This is a good fit for daemons that would fail if there is
+    more than one running per instance, and that can tolerate short
+    periods of unavailability.
+
+    After deployment starts:
+
+    ``` {.shell}
+    $ aws ecs describe-environment-status --environment-name SomeEnvironment/Prod --table
+    REVISION                  STATUS       DESIRED   CURRENT   UP-TO-DATE   AVAILABLE
+    SomeEnvironment/Prod:1    Undeploying  0         5         5            5        
+    SomeEnvironment/Prod:2    Deploying    5         0         0            0        
+    ```
+
+    First task in old revision is terminated:
+
+    ``` {.shell}
+    $ aws ecs describe-environment-status --environment-name SomeEnvironment/Prod --table
+    REVISION                  STATUS       DESIRED   CURRENT   UP-TO-DATE   AVAILABLE
+    SomeEnvironment/Prod:1    Undeploying  0         4         4            4        
+    SomeEnvironment/Prod:2    Deploying    5         0         0            0        
+    ```
+
+    First task in new revision is launched:
+
+    ``` {.shell}
+    $ aws ecs describe-environment-status --environment-name SomeEnvironment/Prod --table
+    REVISION                  STATUS       DESIRED   CURRENT   UP-TO-DATE   AVAILABLE
+    SomeEnvironment/Prod:1    Undeploying  0         4         4            4        
+    SomeEnvironment/Prod:2    Deploying    5         1         1            1        
+    ```
+
+    Deployment progresses in a similar fashion until none of the old
+    revision is left:
+
+    ``` {.shell}
+    $ aws ecs describe-environment-status --environment-name SomeEnvironment/Prod --table
+    REVISION                  STATUS       DESIRED   CURRENT   UP-TO-DATE   AVAILABLE
+    SomeEnvironment/Prod:1    Undeploying  0         0         0            0        
+    SomeEnvironment/Prod:2    Deploying    5         4         4            4        
+    ```
+
+    The final task is launched and the new revision becomes active:
+
+    ``` {.shell}
+    $ aws ecs describe-environment-status --environment-name SomeEnvironment/Prod --table
+    REVISION                  STATUS       DESIRED   CURRENT   UP-TO-DATE   AVAILABLE
+    SomeEnvironment/Prod:1    Inactive     0         0         0            0        
+    SomeEnvironment/Prod:2    Active       5         5         5            5        
+    ```
+
+2.  Terminate after replace
+
```

Comment:
It may be premature at present, but we might want to think about advanced features of this mode.  There are apps out there that do a bunch of management for graceful upgrade, coordinating start of new and hand-off from old via signals (TTOU,TTIN,USR1,etc) and sometimes shared memory. (I'm thinking of things like [haproxy](https://www.haproxy.org/download/1.7/doc/management.txt), [nginx](http://nginx.org/en/docs/control.html), [unicorn](https://bogomips.org/unicorn/SIGNALS.html)). I haven't tried any of this with containers, whose isolation boundaries may make this sorta coordination more difficult.  There's some CLOSED WONTIMPLEMENT docker tickets for [`docker replace`](https://github.com/moby/moby/issues/2733) or other [graceful upgrade capabilities](https://github.com/docker/compose/issues/1663), so it looks like Docker isn't interested in solving the problem, but that doesn't mean we couldn't innovate on behalf of our customers and come up with a workable scheme supported by our scheduler and agent managing the container runtime.